### PR TITLE
[Security] use symfony native way of authentication and session management

### LINKED ISSUE #13876: [Pimcore 11]: Session Management by Symfony (RoadRunner vs FPM)
### Description

I've been playing around with Roadrunner (https://github.com/roadrunner-server/roadrunner) to see if it works with Pimcore and what the Performance is. Thirst thing that didn't work is: the Admin Session is never passed to the client. Reason for that is that the Managed Worker is basically in a CLI sort of Mode:

RoadRunner PHP:
Screenshot 2022-12-22 at 10 31 06

FPM:
Screenshot 2022-12-22 at 10 31 10

Therefore the Session ID is not passed to the Client. (https://www.php.net/manual/en/function.headers-list.php#120539, https://www.php.net/manual/en/features.commandline.differences.php) 

This has to be done within the Application.  Symfony handles that: https://github.com/symfony/symfony/blob/v6.2.0/src/Symfony/Component/HttpKernel/EventListener/AbstractSessionListener.php#L106. But only if the Session has been started. Since Pimcore manages the Session itself like here: https://github.com/pimcore/pimcore/blob/b11f2dd51cf89e92cd27bc5cac1b6830a8a78d01/bundles/AdminBundle/src/Session/Handler/AdminSessionHandler.php#L252 

The session is closed already when the Response is sent and the new SessionID is never passed to the Client.

Pimcore 11 should be refactored in terms of Session Management and it should all be handled by Symfony. I don't see the need for why Pimcore does it.

A simpler solution would be to not close the Session and let Symfony handle that, I tested that and it works for RoadRunner and PHP-FPM.

### PR DESCRIPTION
**Pull request by dpfaffenbauer**
resolves https://github.com/pimcore/pimcore/issues/13876

### CODE REVIEW

File: models/DataObject/Service.php [hunk 1]
```diff
@@ -567,13 +567,18 @@ public static function calculateCellValue(AbstractObject $object, array $helperD
         return null;
     }
 
-    public static function getHelperDefinitions(): mixed
+    public static function getHelperDefinitions(): array
     {
-        return Session::useSession(function (AttributeBagInterface $session) {
-            $existingColumns = $session->get('helpercolumns', []);
+        $stack = \Pimcore::getContainer()->get('request_stack');
+        if ($stack->getMainRequest() && $stack->getMainRequest()->hasSession()) {
```
Comment (jdreesen) [inline models/DataObject/Service.php]:
```suggestion
        if ($stack->getMainRequest()?->hasSession()) {
```

File: bundles/AdminBundle/src/Controller/AdminController.php [hunk 2]
```diff
@@ -92,11 +92,13 @@ public function getTokenResolver()
      */
     protected function getAdminUser(bool $proxyUser = false): User|UserProxy|null
```
Comment (jdreesen) [inline bundles/AdminBundle/src/Controller/AdminController.php:97]: The `$proxyUser` parameter isn't used anymore. Is this intended?

File: lib/Tool/Session.php [hunk 3]
```diff
@@ -11,97 +11,33 @@
...
-use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Session\Attribute\AttributeBagInterface;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
 
 final class Session
 {
-    private static ?AdminSessionHandlerInterface $handler = null;
-
-    public static function getHandler(): AdminSessionHandlerInterface
+    public static function useBag(SessionInterface $session, callable $func, string $namespace = 'pimcore_admin'): mixed
```
Comment (jdreesen) [inline lib/Tool/Session.php:28]:
```suggestion
    /**
     * @param callable(SessionBagInterface, SessionInterface):mixed $func
     */
    public static function useBag(SessionInterface $session, callable $func, string $namespace = 'pimcore_admin'): mixed
```

File: lib/Tool/Session.php [hunk 4]
```diff
@@ -11,97 +11,33 @@
...
-
-    public static function getSessionName(): string
-    {
-        return static::getHandler()->getSessionName();
-    }
-
-    public static function invalidate(): bool
-    {
-        return static::getHandler()->invalidate();
+        return call_user_func_array($func, [$bag, $session]);
```
Comment (jdreesen) [inline lib/Tool/Session.php]: Why not like this?

```suggestion
        return $func($bag, $session);
```

File: lib/Tool/Session.php [hunk 5]
```diff
@@ -11,97 +11,37 @@
...
 namespace Pimcore\Tool;
 
-use Pimcore\Bundle\AdminBundle\Session\Handler\AdminSessionHandler;
-use Pimcore\Bundle\AdminBundle\Session\Handler\AdminSessionHandlerInterface;
-use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Session\Attribute\AttributeBagInterface;
+use Symfony\Component\HttpFoundation\Session\SessionBagInterface;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
 
 final class Session
```
Comment (brusch) [inline lib/Tool/Session.php:23]: @dpfaffenbauer I'm just wondering whether we should really keep this helper class or if we should kick it ... there're no real benefit anymore, right?

Comment (dpfaffenbauer) [reply lib/Tool/Session.php:23]: Since the `getBag` from `SessionInterface` doesn't necessarily return a `AttributeBagInterface`, I thought it is a nice helper to not need to check in every place if the right type is returned.

Comment (brusch) [reply lib/Tool/Session.php:23]: 👍 fine for me, as it also requires less changes in all the other files as a result, but maybe we should tag the class then as `@internal` - then we're flexible in future 😊

Comment (dpfaffenbauer) [reply lib/Tool/Session.php:23]: absolutely

**Comment by brusch** [comment]
Probably resolves #13687 as well 😊

File: bundles/AdminBundle/src/EventListener/AdminAuthenticationDoubleCheckListener.php [hunk 6]
```diff
@@ -1,167 +0,0 @@
...
-use Symfony\Component\HttpKernel\Event\ControllerEvent;
-use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
-use Symfony\Component\HttpKernel\KernelEvents;
-
-/**
- * Handles double authentication check for pimcore controllers after the firewall did to make sure the admin interface is
- * not accessible on configuration errors. Unauthenticated routes are not double-checked (e.g. login).
- *
- * @internal
- */
```
Comment (brusch) [inline bundles/AdminBundle/src/EventListener/AdminAuthenticationDoubleCheckListener.php]: Should we really kick that, it ensures the admin is not accessible even if someone messed up the security config? 🤔

Comment (dpfaffenbauer) [reply bundles/AdminBundle/src/EventListener/AdminAuthenticationDoubleCheckListener.php]: if someone messes up the security config, it is basically their fault.

Comment (fashxp) [reply bundles/AdminBundle/src/EventListener/AdminAuthenticationDoubleCheckListener.php]: ... and still, they will blame Pimcore. I also vote for keeping that.

**Comment by brusch** [comment]
We need to consider that `\Pimcore\Tool\Session` was not internal, so we need to deprecate stuff in `10.6` and probably it also makes sense to provide a BC layer (backports of remaining functions in `Tool\Session`), so that bundles using that functionality can keep compatibility for both pimcore/pimcore v10 and v11. Then of course we could remove the `@internal` again. 

Same for 
`PreAuthenticatedAdminSessionFactory`

File: bundles/AdminBundle/src/EventListener/AdminAuthenticationDoubleCheckListener.php (hunk 6 above)
Comment (dpfaffenbauer) [reply bundles/AdminBundle/src/EventListener/AdminAuthenticationDoubleCheckListener.php]: they will, but it is still their fault. we cannot protect from stupidity. I absolutely don't see any reason that is needed when symfony handles that!

**Comment by brusch** [comment]
@dpfaffenbauer now that https://github.com/pimcore/pimcore/pull/13923 got merged, can you please resolve the conflicts? Thanks a lot and happy new year 🎉🥳

**Comment by dpfaffenbauer** [comment]
@brusch done

Happy new year to you too.

File: bundles/AdminBundle/src/Security/CsrfProtectionHandler.php [hunk 7]
```diff
@@ -45,7 +46,7 @@ public function __construct(array $excludedRoutes, Environment $twig)
 
     public function checkCsrfToken(Request $request)
```
Comment (blankse) [inline bundles/AdminBundle/src/Security/CsrfProtectionHandler.php]:
```suggestion
    public function checkCsrfToken(Request $request): void
```

File: bundles/AdminBundle/src/Security/CsrfProtectionHandler.php [hunk 8]
```diff
@@ -60,21 +61,21 @@ public function checkCsrfToken(Request $request)
...
-                $this->regenerateCsrfToken(false);
+                $this->regenerateCsrfToken($session, false);
             }
         }
 
         return $this->csrfToken;
     }
 
-    public function regenerateCsrfToken(bool $force = true)
+    public function regenerateCsrfToken(SessionInterface $session, bool $force = true)
```
Comment (blankse) [inline bundles/AdminBundle/src/Security/CsrfProtectionHandler.php]:
```suggestion
    public function regenerateCsrfToken(SessionInterface $session, bool $force = true): void
```

File: bundles/AdminBundle/src/Security/CsrfProtectionHandler.php [hunk 9]
```diff
@@ -85,9 +86,9 @@ public function regenerateCsrfToken(bool $force = true)
         $this->twig->addGlobal('csrfToken', $this->csrfToken);
     }
 
-    public function generateCsrfToken()
+    public function generateCsrfToken(SessionInterface $session)
```
Comment (blankse) [inline bundles/AdminBundle/src/Security/CsrfProtectionHandler.php]:
```suggestion
    public function generateCsrfToken(SessionInterface $session): void
```

File: lib/Image/Chromium.php [hunk 10]
```diff
@@ -48,13 +48,15 @@ public static function getChromiumBinary(): ?string
     /**
      * @param string $url
      * @param string $outputFile
+     * @param string $sessionId
+     * @param string $sessionName
      * @param string $windowSize
      *
      * @return bool
      *
```
Comment (blankse) [inline lib/Image/Chromium.php]:
```suggestion
```

File: lib/Tool/Authentication.php [hunk 11]
```diff
@@ -57,26 +68,89 @@ public static function authenticateSession(Request $request = null): ?User
...
-            if (self::isValidUser($user)) {
-                return $user;
-            }
+            return $user instanceof \Pimcore\Bundle\AdminBundle\Security\User\User ? $user->getUser() : null;
         }
 
         return null;
     }
 
+    protected static function safelyUnserialize(string $serializedToken)
```
Comment (blankse) [inline lib/Tool/Authentication.php]:
```suggestion
    protected static function safelyUnserialize(string $serializedToken): mixed
```

File: models/Element/Service.php [hunk 12]
```diff
@@ -1455,7 +1452,7 @@ function (Concrete $object, Data $fieldDefinition, $property, $currentValue) {
      *@internal
      *
      */
-    public static function saveElementToSession(ElementInterface $element, string $postfix = '', bool $clone = true)
+    public static function saveElementToSession(ElementInterface $element, string $sessionId, string $postfix = '', bool $clone = true)
```
Comment (blankse) [inline models/Element/Service.php]:
```suggestion
    public static function saveElementToSession(ElementInterface $element, string $sessionId, string $postfix = '', bool $clone = true): void
```

File: models/Element/Service.php [hunk 13]
```diff
@@ -1494,16 +1491,11 @@ function (Concrete $object, Data $fieldDefinition, $property, $currentValue) {
...
-     * @param string $type
-     * @param int $elementId
-     * @param string $postfix
-     *
-     *@internal
-     *
+     * @internal
      */
-    public static function removeElementFromSession(string $type, int $elementId, string $postfix = '')
+    public static function removeElementFromSession(string $type, int $elementId, string $sessionId, string $postfix = '')
```
Comment (blankse) [inline models/Element/Service.php]:
```suggestion
    public static function removeElementFromSession(string $type, int $elementId, string $sessionId, string $postfix = ''): void
```

**Review by blankse — COMMENTED** [review]
We should add type hints and remove redundant php docs

**Comment by dpfaffenbauer** [comment]
@blankse all done

File: bundles/AdminBundle/src/Controller/Admin/UserController.php [hunk 14]
```diff
@@ -31,10 +31,14 @@
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 use Symfony\Component\HttpFoundation\Session\Attribute\AttributeBagInterface;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
 use Symfony\Component\HttpFoundation\StreamedResponse;
 use Symfony\Component\HttpKernel\Event\ControllerEvent;
 use Symfony\Component\Routing\Annotation\Route;
 use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Core\Validator\Constraints\UserPassword;
+use Symfony\Component\Security\Core\Validator\Constraints\UserPasswordValidator;
```
Comment (jdreesen) [inline bundles/AdminBundle/src/Controller/Admin/UserController.php]: This seems to be unused.

Comment (dpfaffenbauer) [reply bundles/AdminBundle/src/Controller/Admin/UserController.php]: UserPassword is needed, UserPasswordValidator is removed

File: bundles/AdminBundle/src/Controller/Admin/External/AdminerController.php [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+    use Symfony\Component\HttpFoundation\Request;
+    use Symfony\Component\HttpFoundation\Response;
+    use Symfony\Component\HttpKernel\Event\ControllerEvent;
+    use Symfony\Component\HttpKernel\Profiler\Profiler;
+    use Symfony\Component\Routing\Annotation\Route;
+
+    /**
+     * @internal
+     */
+    class AdminerController extends AdminController implements KernelControllerEventInterface
```
Comment (dvesh3) [inline bundles/AdminBundle/src/Controller/Admin/External/AdminerController.php]: please kick it. see https://github.com/pimcore/pimcore/pull/13921

File: bundles/AdminBundle/src/Controller/AdminController.php [hunk 16]
```diff
@@ -144,11 +144,17 @@ public function getTokenResolver()
...
     {
-        if ($proxyUser) {
-            return $this->tokenResolver->getUserProxy();
+        $user = $this->getUser();
+
+        if ($proxyUser && $user instanceof UserProxy) {
```
Comment (dvesh3) [inline bundles/AdminBundle/src/Controller/AdminController.php]: a bit cleaner:
```php

        if (!$user instanceof UserProxy) {
            return null;
        }

        return $proxyUser ? $user : $user->getUser();
```

File: bundles/AdminBundle/src/Controller/Admin/External/AdminerController.php (hunk 15 above)
Comment (dpfaffenbauer) [reply bundles/AdminBundle/src/Controller/Admin/External/AdminerController.php]: @dvesh3 done

File: lib/Tool/Authentication.php [hunk 17]
```diff
@@ -57,26 +68,89 @@ public static function authenticateSession(Request $request = null): ?User
...
+        $session = $request->hasPreviousSession() ? $request->getSession() : null;
+
+        if (null === $session) {
             return null;
         }
 
-        $session = Session::getReadOnly();
-        $user = $session->get('user');
+        $token = $session->get('_security_pimcore_admin');
+        $token = static::safelyUnserialize($token);
```
Comment (dvesh3) [inline lib/Tool/Authentication.php]:
```suggestion
        $token = $token ? static::safelyUnserialize($token) : null;
```
`token` can be null if `authenticateSession` called outside admin session scope. please [see ](https://github.com/pimcore/pimcore/blob/83142b53cfd5a791e3c6167470a66336fd03da22/lib/Targeting/EventListener/ToolbarListener.php#L139)

File: bundles/AdminBundle/config/security_services.yaml [hunk 18]
```diff
@@ -21,39 +22,13 @@ services:
...
             - { name: monolog.logger, channel: security }
 
-    Pimcore\Bundle\AdminBundle\Security\Authenticator\AdminSessionAuthenticator:
-        public: false
-        calls:
-            - [ setLogger, [ '@logger' ] ]
-        tags:
-            - { name: monolog.logger, channel: security }
-
-    pimcore.security.authenticator.admin_pre_auth:
```
Comment (dvesh3) [inline bundles/AdminBundle/config/security_services.yaml:45]: so [authenticate against admin session](https://pimcore.com/docs/pimcore/current/Development_Documentation/Development_Tools_and_Details/Security_Authentication/Authenticate_Against_Admin_Session.html) doesn't work anymore?

File: lib/Tool/Authentication.php [hunk 19]
```diff
@@ -57,26 +68,89 @@ public static function authenticateSession(Request $request = null): ?User
...
-            // renew user
-            $user = User::getById($user->getId());
+        if ($token instanceof TokenInterface) {
+            $token = static::refreshUser($token, \Pimcore::getContainer()->get(UserProvider::class));
+            $user = $token->getUser();
 
-            if (self::isValidUser($user)) {
-                return $user;
-            }
+            return $user instanceof \Pimcore\Bundle\AdminBundle\Security\User\User ? $user->getUser() : null;
```
Comment (dvesh3) [inline lib/Tool/Authentication.php]: I can now login with inactive user credentials. Maybe we should still check a valid user here?

File: bundles/AdminBundle/src/Controller/AdminController.php (hunk 16 above)
Comment (dpfaffenbauer) [reply bundles/AdminBundle/src/Controller/AdminController.php]: done

File: lib/Tool/Authentication.php (hunk 17 above)
Comment (dpfaffenbauer) [reply lib/Tool/Authentication.php]: done

File: bundles/AdminBundle/config/security_services.yaml (hunk 18 above)
Comment (dpfaffenbauer) [reply bundles/AdminBundle/config/security_services.yaml:45]: true, never needed that.... I would rather create the firewall similar to how pimcore's firewall is. Is that actually a used feature?

File: lib/Tool/Authentication.php (hunk 19 above)
Comment (dpfaffenbauer) [reply lib/Tool/Authentication.php]: changed

File: bundles/AdminBundle/config/security_services.yaml (hunk 18 above)
Comment (dvesh3) [reply bundles/AdminBundle/config/security_services.yaml:45]: Honestly, I'm not sure if and how this is being used. Maybe for testing purposes, when editors wants to test some sections on the frontend(which requires authentication) while changing data in the backend. Anyway, if we decide to kick this feature then we need to trigger the deprecation in 10.6.

@brusch do you have any suggestions?

Comment (brusch) [reply bundles/AdminBundle/config/security_services.yaml:45]: @dvesh3 @dpfaffenbauer let's kick that feature - thanks! 😊

Comment (dpfaffenbauer) [reply bundles/AdminBundle/config/security_services.yaml:45]: @brusch ok, I'll deprecate it in 10.6 then

Comment (brusch) [reply bundles/AdminBundle/config/security_services.yaml:45]: Thanks a lot! 👍

File: lib/Tool/Authentication.php (hunk 19 above)
Comment (dvesh3) [reply lib/Tool/Authentication.php]: Now `authenticateSession` returns nulls when user is inactive however i can still login. I guess we are are not validating submitted credentials anywhere in `login_path` or `check_path`.

File: bundles/AdminBundle/src/Security/Authenticator/AdminLoginAuthenticator.php [hunk 20]
```diff
@@ -1,117 +0,0 @@
...
-    public function authenticate(Request $request): Passport
-    {
-        if (!$username = $request->get('username')) {
-            throw new AuthenticationException('Missing username or password');
-        }
-
-        $passport = new Passport(
-            new UserBadge($username),
-            new CustomCredentials(function ($credentials) {
-                $pimcoreUser = Authentication::authenticatePlaintext($credentials['username'], $credentials['password']);
```
Comment (dvesh3) [inline bundles/AdminBundle/src/Security/Authenticator/AdminLoginAuthenticator.php:81]: This authenticator is replaced with `form_login` without moving the logic to validate the user credentials.

File: bundles/AdminBundle/src/Security/Authenticator/AdminSessionAuthenticator.php [hunk 21]
```diff
@@ -1,70 +0,0 @@
...
-     * {@inheritdoc}
-     */
-    public function authenticate(Request $request): Passport
-    {
-        if (!$this->user instanceof UserModel) {
-            throw new AuthenticationException('Invalid User!');
-        }
-
-        $session = Session::getReadOnly();
-        if ($session->has('2fa_required') && $session->get('2fa_required') === true) {
```
Comment (dvesh3) [inline bundles/AdminBundle/src/Security/Authenticator/AdminSessionAuthenticator.php:57]: 2FA authentication doesn't work anymore.

**Comment by dvesh3** [comment]
@dpfaffenbauer please see my comments above. thanks!

**Comment by dvesh3** [comment]
@dpfaffenbauer 🏓

**Comment by dpfaffenbauer** [comment]
@dvesh3 will be done this week. been quite busy the last days

**Comment by dpfaffenbauer** [comment]
@dvesh3 2fa is fixed, I don't know what you mean with the other one?

**Comment by dvesh3** [comment]
> @dvesh3 2fa is fixed, I don't know what you mean with the other one?

@dpfaffenbauer thanks for the fix. About the other one, I mean the provided credentials (username+password) are not validated on admin login anymore. It always authenticates to true with given username.

**Comment by dpfaffenbauer** [comment]
@dvesh3 gotcha, I'll check again

**Comment by dpfaffenbauer** [comment]
@dvesh3 not sure what you are talking about, if I use a wrong password, it doesn't work:

Screenshot 2023-01-23 at 17 23 19

File: lib/Tool/Authentication.php (hunk 19 above)
Comment (dvesh3) [reply lib/Tool/Authentication.php]: This is not resolved with just `authenticateSession` changes. We should update the `Adminbundle\Security\UserProvider::loadUserByIdentifier` with check:
```
$pimcoreUser = PimcoreUser::getByName($identifier);

        if ($pimcoreUser && $pimcoreUser->isActive()) {
```

**Comment by dvesh3** [comment]
> @dvesh3 not sure what you are talking about, if I use a wrong password, it doesn't work:

Sorry, it was only the inactive check which is not working and should be fixed with suggestion here https://github.com/pimcore/pimcore/pull/13909#discussion_r1087737850

**Comment by dpfaffenbauer** [comment]
@dvesh3 OK, I'll check again

File: bundles/AdminBundle/config/pimcore/two_factor_authentication.yaml [hunk 22]
```diff
@@ -1,25 +1,10 @@
-services:
-    _defaults:
-        autowire: true
-        autoconfigure: true
-        public: false
-
-    Pimcore\Bundle\AdminBundle\EventListener\TwoFactorListener:
```
Comment (dvesh3) [inline bundles/AdminBundle/config/pimcore/two_factor_authentication.yaml:7]: we still need to register the event listener for `onAuthenticationAttempt`..right?

**Comment by dpfaffenbauer** [comment]
@dvesh3 2fa is fixed and inactive check is also there now

File: bundles/AdminBundle/config/pimcore/two_factor_authentication.yaml (hunk 22 above)
Comment (dpfaffenbauer) [reply bundles/AdminBundle/config/pimcore/two_factor_authentication.yaml:7]: is not needed, I refactored the 2fa a bit. I am still not happy with how it is implemented. If 2fa is required, you still get logged into the backend and can remove the Overlay with 2fa or even the popup and still work in Pimcore. I tried to refactor to have a separate login screen, but it is not easy to implement... the implementation is quite tricky. I think we should leave it as is and create another PR to make that better.

**Comment by dpfaffenbauer** [comment]
@dvesh3 I refactored the 2fa setup process now too. It works without actually loading the full backend

**Comment by dpfaffenbauer** [comment]
rebased too fix merge issues

File: bundles/AdminBundle/src/Controller/Admin/Asset/AssetController.php [hunk 23]
```diff
@@ -426,7 +428,7 @@ public function existsAction(Request $request): JsonResponse
      * @param Request $request
      * @param Config $config
      *
-     * @return array{success: bool, asset: ?Asset}
```
Comment (jdreesen) [inline bundles/AdminBundle/src/Controller/Admin/Asset/AssetController.php]: Why did you remove this? It's still true, right? 
The changes in this file look to me like a conflict resolution gone wrong.
@blankse can you have a look, please?

File: bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectController.php [hunk 24]
```diff
@@ -891,14 +970,14 @@ public function addAction(Request $request, Model\FactoryInterface $modelFactory
...
             $parent = DataObject\Concrete::getById($parentId);
             $object->setClassId($parent->getClass()->getId());
         }
 
-        $object->setClassName($request->request->get('className'));
-        $object->setParentId($request->request->getInt('parentId'));
-        $object->setKey($request->request->get('key'));
+        $object->setClassName($request->get('className'));
+        $object->setParentId($request->get('parentId'));
+        $object->setKey($request->get('key'));
```
Comment (jdreesen) [inline bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectController.php]: Same as above.

File: bundles/AdminBundle/src/Controller/Admin/ElementController.php [hunk 25]
```diff
@@ -339,10 +339,10 @@ public function noteAddAction(Request $request): JsonResponse
...
     {
         $element = null;
-        if ($request->query->get('id')) {
-            $element = Element\Service::getElementById($request->query->get('type'), $request->query->getInt('id'));
-        } elseif ($request->query->get('path')) {
-            $element = Element\Service::getElementByPath($request->query->get('type'), $request->query->get('path'));
+        if ($request->get('id')) {
+            $element = Element\Service::getElementById($request->get('type'), $request->get('id'));
+        } elseif ($request->get('path')) {
+            $element = Element\Service::getElementByPath($request->get('type'), $request->get('path'));
```
Comment (jdreesen) [inline bundles/AdminBundle/src/Controller/Admin/ElementController.php]: Apparently, it seems to be the case for many files. Looks to me like the "use my version" conflict resolution strategy was used.

File: models/Element/Service.php [hunk 26]
```diff
@@ -236,15 +245,22 @@ private static function getDependencyForFrontend(ElementInterface $element): arr
...
-        };
+        if ($config['type'] == 'object') {
+            return DataObject::getById($config['id']);
+        } elseif ($config['type'] == 'asset') {
+            return Asset::getById($config['id']);
+        } elseif ($config['type'] == 'document') {
+            return Document::getById($config['id']);
+        }
+
+        return null;
```
Comment (jdreesen) [inline models/Element/Service.php]: Here too. This reverts the changes from #13997.

**Review by jdreesen — CHANGES_REQUESTED** [review]

**Comment by dpfaffenbauer** [comment]
@jdreesen It was a messy re-base and I messed it up. I manually fixed it now

File: models/Element/Service.php [hunk 27]
```diff
@@ -391,7 +386,8 @@ public static function getBaseClassNameForElement(string|ElementInterface $eleme
      *
      * @return string
      *
-     * @deprecated will be removed in Pimcore 11, use getSafeCopyName() instead
+     *@deprecated will be removed in Pimcore 11, use getSafeCopyName() instead
```
Comment (jdreesen) [inline models/Element/Service.php]:
```suggestion
     * @deprecated will be removed in Pimcore 11, use getSafeCopyName() instead
```

File: models/Element/Service.php [hunk 28]
```diff
@@ -1422,21 +1406,15 @@ function (Concrete $object, Data $fieldDefinition, $property, $currentValue) {
...
     /**
-     * @param string $type
-     * @param int $elementId
-     * @param string $postfix
-     *
-     * @internal
-     */
-    public static function removeElementFromSession(string $type, int $elementId, string $postfix = ''): void
+     * @internal/
+    public static function removeElementFromSession(string $type, int $elementId, string $sessionId, string $postfix = ''): void
```
Comment (jdreesen) [inline models/Element/Service.php]:
```suggestion
     * @internal
     */
    public static function removeElementFromSession(string $type, int $elementId, string $sessionId, string $postfix = ''): void
```

**Review by jdreesen — COMMENTED** [review]
Looks much better now :)

File: models/Element/Service.php (hunk 28 above)
Comment (dpfaffenbauer) [reply models/Element/Service.php]: just found that one too ;)

File: bundles/AdminBundle/config/pimcore/two_factor_authentication.yaml (hunk 22 above)
Comment (dvesh3) [reply bundles/AdminBundle/config/pimcore/two_factor_authentication.yaml:7]: see https://github.com/pimcore/pimcore/pull/13909#issuecomment-1406683977

**Review by dvesh3 — APPROVED** [review]

**Comment by dpfaffenbauer** [comment]
@dvesh3 lets merge it ;) and I get my first Pimcore Partner Stars rewarded :P

**Comment by dvesh3** [comment]
@dpfaffenbauer soon! good things take time 😉

File: lib/Tool/Authentication.php [hunk 29]
```diff
@@ -57,26 +62,91 @@ public static function authenticateSession(Request $request = null): ?User
...
-        $session = Session::getReadOnly();
-        $user = $session->get('user');
+        $token = $session->get('_security_pimcore_admin');
+        $token = $token ? static::safelyUnserialize($token) : null;
 
-        if ($user instanceof User) {
-            // renew user
-            $user = User::getById($user->getId());
+        if ($token instanceof TokenInterface) {
+            $token = static::refreshUser($token, \Pimcore::getContainer()->get(UserProvider::class));
```
Comment (brusch) [inline lib/Tool/Authentication.php:75]: This feels a bit odd to me, to be honest. 
Shouldn't we try to the he user from `Symfony\Bundle\SecurityBundle\Security` or `security.token_storage` ?

**Review by brusch — CHANGES_REQUESTED** [review]

File: lib/Tool/Authentication.php (hunk 29 above)
Comment (dpfaffenbauer) [reply lib/Tool/Authentication.php:75]: reason for that is, that we can authenticate against the Admin Session from outside the firewall. It was there before, so I kept it.

Comment (dpfaffenbauer) [reply lib/Tool/Authentication.php:75]: I think this is needed to disable frontend cache (for example) when logged into admin

Comment (brusch) [reply lib/Tool/Authentication.php:75]: Mhm, I see - and there's no other way to get this information outside the firewall (pattern)?

Comment (dpfaffenbauer) [reply lib/Tool/Authentication.php:75]: Not that I am aware of.

Comment (brusch) [reply lib/Tool/Authentication.php:75]: What about trying to use `\Symfony\Component\Security\Http\Firewall\ContextListener::authenticate()` with a dummy request going to an admin route, but containing the real session? 

I'm a bit afraid, that something might change how Symfony deals with tokens and then this would completely break. At least we should give it a try 🤔

**Review by brusch — CHANGES_REQUESTED** [review]

File: lib/Tool/Authentication.php (hunk 29 above)
Comment (dpfaffenbauer) [reply lib/Tool/Authentication.php:75]: @brusch I am afraid that this is also not going to work, the ContextListener would set the token into the token storage, therefore overwriting the current token. So if we use that, and the user is logged into another firewall, this check would overwrite the token.

Comment (brusch) [reply lib/Tool/Authentication.php:75]: Hmm, good point. I guess then it's fine if we start that way and we create a follow-up ticket for checking out alternatives, or maybe discuss them with the guys from Symfony.
